src/App.tsx

```diff
@@ -1,26 +1,266 @@
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState(Filter.All);
+  const [tempTodoTitle, setTempTodoTitle] = useState<string | null>('');
+  const [idsProccesing, setIdsProccesing] = useState<number[]>([]);
+
+  const ref = useRef<HTMLInputElement | null>(null);
+
+  const handleAddTodo = async (title: string) => {
+    const formattedTitle = title.trim();
+
+    if (!formattedTitle) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    const newTodo = {
+      title: formattedTitle,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    try {
+      setTempTodoTitle(title); // Show the temporary todo
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(currentTodos => [...currentTodos, createdTodo]);
+    } catch {
+      setErrorMessage('Unable to add a todo');
+      throw new Error('Unable to add a todo');
+    } finally {
+      setTempTodoTitle(null);
+    }
+  };
+
+  const handleDeleteTodo = async (id: number) => {
+    try {
+      setIdsProccesing([id]);
+      await deleteTodo(id);
+
+      setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
+    } catch {
+      setErrorMessage('Unable to delete a todo');
+    } finally {
+      setIdsProccesing([]);
+    }
+  };
+
+  const handleEditTodo = async (id: number, data: Partial<Todo>) => {
+    try {
+      setIdsProccesing([id]);
+      const editedTodo = await patchTodo(id, data);
+
+      setTodos(currentTodos =>
+        currentTodos.map(todo => {
+          if (todo.id === id) {
+            return editedTodo;
+          }
+
+          return todo;
+        }),
+      );
+    } catch {
+      setErrorMessage('Unable to update a todo');
+      throw new Error('Unable to update a todo');
+    } finally {
+      setIdsProccesing([]);
+    }
+  };
+
+  const clearCompletedTodos = async () => {
+    const filteredTodos = todos.filter(todo => todo.completed);
+    const completedIds = filteredTodos.map(todo => todo.id);
+
+    setIdsProccesing(completedIds);
+
+    try {
+      const deleteCallback = async (todo: Todo) => {
+        try {
+          await deleteTodo(todo.id);
+
+          return { id: todo.id, status: 'resolved' };
+        } catch {
+          setErrorMessage('Unable to delete a todo');
+
+          return { id: todo.id, status: 'rejected' };
+        } finally {
+          setIdsProccesing([]);
+        }
+      };
+
+      const res = await Promise.allSettled(filteredTodos.map(deleteCallback));
+
+      const resolvedIds = res.reduce(
+        (acc, item) => {
+          if (item.status === 'rejected') {
+            return acc;
+          }
+
+          if (item.value.status === 'resolved') {
+            return { ...acc, [item.value.id]: item.value.id };
```

Comment:
```suggestion
            return { ...acc, id: item.value.id };
```
maybe it will be better